c++/src/kj/compat/http.c++

```diff
@@ -6177,74 +6291,115 @@ private:
+      kj::Own<kj::AsyncIoStream> guarded;
+      kj::Own<kj::PromiseFulfiller<void>> fulfiller;
+    };
 
-      auto wrapper = kj::heap<DelayedEofInputStream>(
-          kj::mv(pipe.in), task.attach(kj::addRef(*this)));
+    kj::Own<kj::PromiseFulfiller<HttpClient::ConnectRequest::Status>> fulfiller;
+    StreamsAndFulfiller streamAndFulfiller;
+    bool connectStreamDetached = false;
+
+    StreamsAndFulfiller initStreamsAndFulfiller(kj::Own<kj::AsyncIoStream> stream) {
+      auto paf = kj::newPromiseAndFulfiller<void>();
+      auto guarded = kj::heap<AsyncIoStreamWithGuards>(
+          kj::mv(stream),
+          kj::Maybe<HttpInputStreamImpl::ReleasedBuffer>(nullptr),
+          kj::mv(paf.promise));
+      return StreamsAndFulfiller {
+        kj::mv(guarded),
+        kj::mv(paf.fulfiller)
+      };
+    }
 
-      task = task.then([this,statusCode,statusTextCopy = kj::mv(statusTextCopy),
-                        headersCopy = kj::mv(headersCopy), wrapper = kj::mv(wrapper)]() mutable {
-        kj::Own<kj::AsyncInputStream> body =
-            wrapper.attach(kj::mv(statusTextCopy), kj::mv(headersCopy));
-        fulfiller->fulfill(HttpClient::ConnectResponse {
-          statusCode,
-          statusTextCopy,
-          headersCopy.get(),
-          kj::mv(body),
-        });
-      }).eagerlyEvaluate([](kj::Exception&& e) { KJ_LOG(ERROR, e); });
+    void handleException(kj::Exception&& ex, kj::Own<kj::AsyncIoStream> connectStream) {
+      // Log the exception...
+      KJ_LOG(ERROR, "Error in HttpClientAdapter connect()", kj::cp(ex));
+      // Reject the status promise if it is still pending...
+      if (fulfiller->isWaiting()) {
+        fulfiller->reject(kj::cp(ex));
+      }
+      if (streamAndFulfiller.fulfiller->isWaiting()) {
+        // If the guard hasn't yet ben released, we can fail the pending reads by
+        // rejecting the fulfiller here.
+        streamAndFulfiller.fulfiller->reject(kj::mv(ex));
+      } else {
+        // The guard has already been released at this point.
+        // TODO(connect): How to properly propagate the actual exception to the
+        // connect stream? Here we "simply" shut it down.
+        connectStream->abortRead();
+        connectStream->shutdownWrite();
+      }
+    }
 
-      return kj::mv(pipe.out);
+    kj::Own<kj::AsyncIoStream> getConnectStream() {
+      KJ_ASSERT(!connectStreamDetached, "the connect stream was already detached");
+      connectStreamDetached = true;
+      return streamAndFulfiller.guarded.attach(kj::addRef(*this));
     }
 
-  private:
-    kj::Own<kj::PromiseFulfiller<HttpClient::ConnectResponse>> fulfiller;
-    kj::Promise<void> task = nullptr;
+    void respond(uint statusCode,
+                 kj::StringPtr statusText,
+                 const HttpHeaders& headers,
+                 kj::Maybe<kj::Own<kj::AsyncInputStream>> errorBody = nullptr) {
+      KJ_REQUIRE(fulfiller->isWaiting(), "the connect request already has a response");
```

Comment:
This is not an appropriate way to check if the fulfiller was already fulfilled -- this will also fail if the associated promise has been dropped (and thus is no longer waiting on the fulfiller). We should remove this line.

Reply:
Nice find! Can confirm, this resolves the issue I was seeing